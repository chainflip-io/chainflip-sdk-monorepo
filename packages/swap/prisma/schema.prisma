// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
