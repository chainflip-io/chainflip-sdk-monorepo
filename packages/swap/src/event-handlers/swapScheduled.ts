// Set the column in the DB to the block timestamp and the deposit amount.
import assert from 'assert';
import { z } from 'zod';
import { stateChainAssetEnum, unsignedInteger } from '@/shared/parsers';
import logger from '../utils/logger';
import { encodedAddress, foreignChainAddress } from './common';
import type { EventHandlerArgs } from '.';

const baseArgsWithoutSource = z.object({
  swapId: unsignedInteger,
  depositAmount: unsignedInteger,
  destinationAsset: stateChainAssetEnum,
  destinationAddress: encodedAddress,
});

// TODO:0.9 remove this
const depositAssetArg = z
  .object({
    depositAsset: stateChainAssetEnum,
  })
  .transform(({ depositAsset }) => ({ sourceAsset: depositAsset }));

const sourceAssetArg = z.object({
  sourceAsset: stateChainAssetEnum,
});

const baseArgs = z.union([
  z.intersection(baseArgsWithoutSource, depositAssetArg),
  z.intersection(baseArgsWithoutSource, sourceAssetArg),
]);

const depositChannelOrigin = z.object({
  __kind: z.literal('DepositChannel'),
  // TODO 0.9: depositAddress is a "EncodedAddress" in 0.9: https://github.com/chainflip-io/chainflip-backend/pull/3394
  depositAddress: z.union([foreignChainAddress, encodedAddress]),
});

const vaultOrigin = z.object({
  __kind: z.literal('Vault'),
  txHash: z.string(),
});

const eventArgs = z.intersection(
  baseArgs,
  z.object({ origin: z.union([depositChannelOrigin, vaultOrigin]) }),
);

export type SwapScheduledEvent = z.input<typeof eventArgs>;

export default async function swapScheduled({
  prisma,
  block,
  event,
}: EventHandlerArgs): Promise<void> {
  try {
    const { swapId, depositAmount, ...args } = eventArgs.parse(event.args);

    const newSwapData = {
      depositReceivedBlockIndex: `${block.height}-${event.indexInBlock}`,
      depositAmount: depositAmount.toString(),
      nativeId: swapId,
      depositReceivedAt: new Date(block.timestamp),
    };

    if (args.origin.__kind === 'DepositChannel') {
      const depositAddress = args.origin.depositAddress.address;

      const channels = await prisma.swapDepositChannel.findMany({
        where: {
          srcAsset: args.sourceAsset,
          depositAddress,
          expiryBlock: { gte: block.height },
        },
      });

      if (channels.length === 0) {
        logger.info(
          `SwapScheduled: SwapDepositChannel not found for depositAddress ${depositAddress}`,
        );
        return;
      }

      assert(
        channels.length === 1,
        `SwapScheduled: too many active swap intents found for depositAddress ${depositAddress}`,
      );

      const [{ srcAsset, destAddress, destAsset, id }] = channels;

      await prisma.swap.create({
        data: {
          swapDepositChannelId: id,
          srcAsset,
          destAsset,
          destAddress,
          ...newSwapData,
        },
      });
    } else if (args.origin.__kind === 'Vault') {
      await prisma.swap.create({
        data: {
          srcAsset: args.sourceAsset,
          destAsset: args.destinationAsset,
          destAddress: args.destinationAddress.address,
          ...newSwapData,
        },
      });
    }
  } catch (error) {
    logger.customError(
      'error in "SwapScheduled" handler',
      { alertCode: 'EventHandlerError' },
      { error, handler: 'SwapScheduled' },
    );
    throw error;
  }
}
