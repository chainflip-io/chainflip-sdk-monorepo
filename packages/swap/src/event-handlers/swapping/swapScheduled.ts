import { swappingSwapScheduled } from '@chainflip/processor/180/swapping/swapScheduled';
import { z } from 'zod';
import type { EventHandlerArgs } from '..';

export type SwapScheduledArgs = z.input<typeof swappingSwapScheduled>;

const swapTypeMap = {
  CcmGas: 'GAS',
  CcmPrincipal: 'PRINCIPAL',
  Swap: 'SWAP',
  NetworkFee: 'NETWORK_FEE',
  IngressEgressFee: 'INGRESS_EGRESS_FEE',
} as const;

export default async function swapScheduled({
  prisma,
  block,
  event,
}: EventHandlerArgs): Promise<void> {
  const { swapId, swapRequestId, inputAmount, swapType } = swappingSwapScheduled.parse(event.args);

  const swapRequest = await prisma.swapRequest.findUniqueOrThrow({
    where: { nativeId: swapRequestId },
  });

  await prisma.swap.create({
    data: {
      swapRequestId: swapRequest.id,
      swapInputAmount: inputAmount.toString(),
      srcAsset: swapRequest.srcAsset,
      destAsset: swapRequest.destAsset,
      nativeId: swapId,
      type: swapTypeMap[swapType],
      swapScheduledAt: new Date(block.timestamp),
      swapScheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
    },
  });
}
