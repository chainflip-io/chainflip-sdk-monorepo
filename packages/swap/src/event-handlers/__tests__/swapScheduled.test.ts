import prisma, { SwapDepositChannel } from '../../client';
import swapScheduled from '../swapScheduled';
import {
  createDepositChannel,
  swapScheduledDepositChannelMock,
  swapScheduledVaultMock,
} from './utils';

describe(swapScheduled, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  describe('deposit channel origin', () => {
    let swapDepositChannel: SwapDepositChannel;

    beforeEach(async () => {
      swapDepositChannel = await createDepositChannel({
        srcAsset:
          swapScheduledDepositChannelMock.eventContext.event.args.sourceAsset
            .__kind,
        destAsset:
          swapScheduledDepositChannelMock.eventContext.event.args
            .destinationAsset.__kind,
        depositAddress:
          swapScheduledDepositChannelMock.eventContext.event.args.origin
            .depositAddress.value,
        destAddress:
          swapScheduledDepositChannelMock.eventContext.event.args
            .destinationAddress.value,
      });
    });

    it('stores a new swap from a deposit channel', async () => {
      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDepositChannelMock.block as any,
          event: swapScheduledDepositChannelMock.eventContext.event as any,
        });
      });

      const swap = await prisma.swap.findFirstOrThrow({
        where: { swapDepositChannelId: swapDepositChannel.id },
      });

      expect(swap.depositAmount.toString()).toEqual(
        swapScheduledDepositChannelMock.eventContext.event.args.depositAmount,
      );
      expect(swap).toMatchSnapshot({
        id: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        swapDepositChannelId: expect.any(BigInt),
      });
    });

    it('does not store a new swap if the deposit channel is expired', async () => {
      await prisma.swapDepositChannel.update({
        where: { id: swapDepositChannel.id },
        data: { expiryBlock: -1 },
      });

      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDepositChannelMock.block,
          event: swapScheduledDepositChannelMock.eventContext.event as any,
        });
      });

      expect(await prisma.swap.findFirst()).toBeNull();
    });

    it('does not store a new swap if the deposit channel is not found', async () => {
      await prisma.swapDepositChannel.update({
        where: { id: swapDepositChannel.id },
        data: { depositAddress: '0x0' },
      });

      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDepositChannelMock.block,
          event: swapScheduledDepositChannelMock.eventContext.event as any,
        });
      });

      expect(await prisma.swap.findFirst()).toBeNull();
    });

    it('does not store a new swap if the deposit channel is not unique', async () => {
      const { id, uuid, ...rest } = swapDepositChannel;

      await prisma.swapDepositChannel.create({ data: rest });

      await expect(
        prisma.$transaction(async (client) => {
          await swapScheduled({
            prisma: client,
            block: swapScheduledDepositChannelMock.block,
            event: swapScheduledDepositChannelMock.eventContext.event as any,
          });
        }),
      ).rejects.toThrowError();

      expect(await prisma.swap.findFirst()).toBeNull();
    });
  });

  describe('smart contract origin', () => {
    it('stores a new swap from a contract deposit', async () => {
      // create a swap after receiving the event
      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledVaultMock.block as any,
          event: swapScheduledVaultMock.eventContext.event as any,
        });
      });

      const swap = await prisma.swap.findFirstOrThrow();

      expect(swap.depositAmount.toString()).toEqual(
        swapScheduledVaultMock.eventContext.event.args.depositAmount,
      );
      expect(swap).toMatchSnapshot({
        id: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });
    });
  });
});
