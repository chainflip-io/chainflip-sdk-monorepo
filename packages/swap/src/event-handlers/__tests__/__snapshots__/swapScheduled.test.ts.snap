// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapScheduled deposit channel origin stores a new swap from a deposit channel 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "222222222222222222",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "0x6AA69332b63BB5B1d7CA5355387edd5624e181f3",
  "destAsset": "Flip",
  "egressCompletedAt": null,
  "egressCompletedBlockIndex": null,
  "id": Any<BigInt>,
  "nativeId": 9876545n,
  "srcAsset": "Eth",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "updatedAt": Any<Date>,
}
`;

exports[`swapScheduled smart contract origin stores a new swap from a contract deposit 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "222222222222222222",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
  "destAsset": "FLIP",
  "egressCompletedAt": null,
  "egressCompletedBlockIndex": null,
  "id": Any<BigInt>,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": null,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "updatedAt": Any<Date>,
}
`;
